pkg/operator/targetconfigcontroller/targetconfigcontroller.go

```diff
@@ -149,3 +162,38 @@ func (c TargetConfigController) sync(ctx context.Context, syncContext factory.Sy
 	)
 	return err
 }
+
+func (c TargetConfigController) applyClusterCSIDriverChange(
+	infra *ocpconfigv1.Infrastructure, sourceCFG vsphere.VSphereConfig, clusterCSIDriver *opv1.ClusterCSIDriver) (*corev1.ConfigMap, error) {
+	csiConfigString := string(c.csiConfigManifest)
+
+	for pattern, value := range map[string]string{
+		"${CLUSTER_ID}":  infra.Status.InfrastructureName,
+		"${VCENTER}":     sourceCFG.Workspace.VCenterIP,
+		"${DATACENTERS}": sourceCFG.Workspace.Datacenter, // TODO: datacenters?
+	} {
+		csiConfigString = strings.ReplaceAll(csiConfigString, pattern, value)
+	}
+
+	csiConfig, err := iniv1.Load([]byte(csiConfigString))
+	if err != nil {
+		return nil, fmt.Errorf("error loading ini file: %v", err)
+	}
+
+	driverConfig := clusterCSIDriver.Spec.DriverConfig
```

Comment:
I don't know. I could try but yeah driver crashing with empty value of `topology-categories` was exactly what I was worried about and hence I chose to parse and load the ini file separately.

Reply:
Also - my motivation was to reduce amount of breakage that can happen from future changes. Who is to say, if driver handles empty topology now, it won't break in future. :-) So to be on safer side..

Reply:
So the only concern is the use of empty values, and so it's more safe to have `cloud.conf: ""` by default? What if only some of the values in cloud conf are empty? Should it be an error or will it work fine? I see few potential issues in the current version:

1. ` applyClusterCSIDriverChange()` does not validate the values before doing `.ReplaceAll()` so it does not avoid empty strings (maybe it should have it regardless of ini vs template approach?)

2.  the resulting config map is obscured, it's very diffucult to tell what it will provide under `data:` and the label is hidden completely. If we go with a template instead we could use a control variable and conditions, something like this would be more transparent and robust:

```
apiVersion: v1
kind: ConfigMap
metadata:
  namespace: openshift-cluster-csi-drivers
  name: vsphere-csi-config
  {{ - if .Values.ValidConfig }}
  labels: 
    "topology-categories": {{ .Values.TopologyCategories }}
  {{ - end }}
data:
{{ - if .Values.ValidConfig }}
  [Global]
  cluster-id = {{ .Values.Id }}
  [VirtualCenter {{ .Values.VCenter }}]
  insecure-flag = "true"
  datacenters = {{ .Values.Datacenters }}
{{ else }}
  cloud.conf: ""
{{ - end }}
```

Reply:
>The resulting config map is obscured, it's very diffucult to tell what it will provide under data: and the label is hidden completely.

hmm, I am not sure I like the templatized YAML file. It is also slightly wrong btw. The `topology-categories` has to be in `data` field and part of ini content, not the configmap label. Anyways - do you think this is cleaner? I am not so sure, I have a long mistrust of templatizing YAMLs in a language that is not whitespace sensitive (like go). It is too easy to make a mistake and get indentation wrong.

Reply:
I see, I misinterpreted [this](https://github.com/openshift/vmware-vsphere-csi-driver-operator/pull/107/files#diff-0af15f182aea9898e784fc589175ef4c2411c9391bb5a1f41327ba7c3e2172f5R182) line. No need to go this route, it just seemed like we need to do a very similar thing to what Helm does. Surely it's prone to indentation errors, but it might be a good tradeoff for the advantages.

If we're quite certain this will not grow and more logic will be needed around generating config maps (or other resources) it's probably fine to go with ini and string replacement.

Reply:
I am not a big fan of huge templates either. IMO it's enough if we add comments to assets that certain fields are set by the operator.